### PR TITLE
Implement queue.copy_external_image_to_texture for WebGL2 and improve WebGPU Impl

### PR DESCRIPTION
**Pull request by cwfitzgerald**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

Still todo:
- [x] Merge https://github.com/grovesNL/glow/pull/234
- [x] Add tests

**Connections**

Closes https://github.com/gfx-rs/wgpu/issues/1888.

Follows up on #2781.

**Description**

Implements copy_external_image_to_texture fully on both WebGPU and WebGL - including all 4 different types that are supported by WebGPU.

**Testing**

Testing todo.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -57,6 +57,7 @@ env_logger = "0.9"
 futures-intrusive = "0.4"
 fxhash = "0.2.1"
 glam = "0.21.3"
+image = { version = "0.24", default-features = false, features = ["png"] }
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply Cargo.toml:60]: This is only used by `wgpu`'s tests as a dev dependency

**Review by teoxoy — COMMENTED** [review]
Have a few comments, otherwise it looks good!

File: wgpu-types/src/lib.rs [hunk 2]
```diff
@@ -4091,10 +4102,40 @@ pub enum TextureDimension {
...
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "trace", derive(Serialize))]
+#[cfg_attr(feature = "replay", derive(Deserialize))]
+#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
+pub struct Origin2d {
+    ///
+    pub x: u32,
+    ///
+    pub y: u32,
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-types/src/lib.rs:4118]: They were missing on the Extent2D struct too. I'm not too sure what I would even write in there non-redundantly.

I think these were just there to allow turning on the missing docs lint.

File: wgpu/tests/external_texture.rs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+                        // It's now out bounds for the destination
+                        dest_width = 2;
+                    }
+                    TestCase::MultiSliceCopy => {
+                        valid = false;
+                        copy_size.depth_or_array_layers = 2;
+                        dest_layers = 2;
+                    }
+                    TestCase::SecondSliceCopy => {
+                        correct = false; // TODO: what?
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu/tests/external_texture.rs:249]: Correct - will file an issue.

File: wgpu/src/backend/web.rs [hunk 4]
```diff
@@ -471,12 +477,24 @@ fn map_texture_copy_view(view: crate::ImageCopyTexture) -> web_sys::GpuImageCopy
...
 fn map_tagged_texture_copy_view(
-    view: crate::ImageCopyTexture,
+    view: crate::ImageCopyTextureTagged,
 ) -> web_sys::GpuImageCopyTextureTagged {
     let texture = &<<Context as crate::Context>::TextureId>::from(view.texture.id).0;
     let mut mapped = web_sys::GpuImageCopyTextureTagged::new(texture);
     mapped.mip_level(view.mip_level);
     mapped.origin(&map_origin_3d(view.origin));
+    mapped.aspect(map_texture_aspect(view.aspect));
+    // mapped.color_space(map_color_space(view.color_space));
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu/src/backend/web.rs:487]: Yes - will file

File: wgpu/src/backend/web.rs [hunk 5]
```diff
@@ -412,6 +412,12 @@ fn map_vertex_step_mode(mode: wgt::VertexStepMode) -> web_sys::GpuVertexStepMode
     }
 }
 
+// fn map_extent_2d(extent: wgt::Extent2d) -> web_sys::GpuExtent2dDict {
+//     let mut mapped = web_sys::GpuExtent2dDict::new(extent.width);
+//     mapped.height(extent.height);
+//     mapped
+// }
+
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu/src/backend/web.rs:420]: This is needed for https://github.com/gfx-rs/wgpu/pull/3288#discussion_r1084016443 - will uncomment as it's now used.

File: wgpu-types/src/lib.rs [hunk 6]
```diff
@@ -1139,6 +1139,17 @@ bitflags::bitflags! {
...
         /// Corresponds to Vulkan's `VkPhysicalDeviceFeatures.depthBiasClamp`
         const DEPTH_BIAS_CLAMP = 1 << 18;
+
+        /// With this feature not present, there are the following restrictions on `Queue::copy_external_image_to_texture`:
+        /// - The source must not be [`web_sys::OffscreenCanvas`]
+        /// - [`ImageCopyExternalImage::origin`] must be zero.
+        /// - [`ImageCopyTextureTagged::color_space`] must be srgb.
+        /// - If the source is an [`web_sys::ImageBitmap`]:
+        ///   - [`ImageCopyExternalImage::flip_y`] must be false.
+        ///   - [`ImageCopyTextureTagged::premultiplied_alpha`] must be false.
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-types/src/lib.rs:1149]: Yup - I get why, but it's annoying.

File: wgpu/src/backend/web.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu/src/backend/web.rs:420]: Got complexly flummoxed - this was supposed to be map_origin_2d. Fixed

File: wgpu-hal/src/gles/command.rs [hunk 7]
```diff
@@ -309,6 +309,32 @@ impl crate::CommandEncoder<super::Api> for super::CommandEncoder {
...
+        src: &wgt::ImageCopyExternalImage,
+        dst: &super::Texture,
+        dst_premultiplication: bool,
+        regions: T,
+    ) where
+        T: Iterator<Item = crate::TextureCopy>,
+    {
+        let (dst_raw, dst_target) = dst.inner.as_native();
+        for copy in regions {
+            // TODO: Clamp size?
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-hal/src/gles/command.rs:324]: Yeah.

**Review by teoxoy — APPROVED** [review]
Good stuff!

**Pull request merged**